### ISSUE TITLE
homectl: subcommand to restore a missing or corrupt .identity file inside encrypted home directory

### ISSUE THREAD

**Issue by zackw** [systemd/systemd#17688]
**Is your feature request related to a problem? Please describe.**

The `.identity` file in a homed-managed home directory is not supposed to be moved or modified by the user whose home directory it is, but nothing prevents them from doing so (and AFAIK prevention is not possible).  If they do move or modify the file, and then log out, they won't be able to log back in until the file is restored somehow.

If the home directory is not encrypted, root can copy a known-good version of the identity file from `/var/lib/systemd/home` into the storage area, but if the home directory *is* encrypted, it would need to be decrypted first.  But `homectl activate` will fail, for the same reason logging in as the user fails.  Depending on which encrypted backing store is in use, it may be very difficult to decrypt the directory by hand (for instance, for fscrypt directories, `homectl create` does not set up metadata compatible with the `fscrypt` utility).

**Describe the solution you'd like**

Please add a subcommand to `homectl` that prompts for authentication, like `activate` does, and if that succeeds, decrypts the storage area and restores the identity file from the known-good copy in `/var/lib/systemd/home`.  Any existing `.identity` file should be renamed out of the way, to `.identity.corrupt` or some such.

**Describe alternatives you've considered**

Alternatively, `homectl activate` could grow an option to ignore the `.identity` file (and/or the absence of an `.identity` file) inside the storage area, thus allowing root to access the storage area (as long as they can authenticate as the user) and re-create the identity file manually.  This might be useful for other troubleshooting purposes, but the simple scenario described above would need more steps to complete recovery.

**Comment by poettering** [comment]
(btw, you can decrypt the home dir manually, and insert the identity file again. It's all plain standard stuff, we carefuly make sure you can use the usual tools to access the home dir, if you do not have homed around)

**Comment by zackw** [comment]
(I tried, but could not figure out how to decrypt a directory with an fscrypt backing store.  Like I said, it does not set up metadata compatible with the standalone `fscrypt` utility.  Perhaps the procedure for manually decrypting each type of encrypted home directory should be documented.)

**Comment by poettering** [comment]
yeah, iirc fscrypt uses a borked KDF. They implemented it themselves, and really shouldn't have. We use OpenSSL. This means the hashes don't match even though on paper we implement the same stuff.

**Comment by ebiggers** [comment]
> yeah, iirc fscrypt uses a borked KDF. They implemented it themselves, and really shouldn't have. We use OpenSSL. This means the hashes don't match even though on paper we implement the same stuff.

I'm not sure what you're talking about here.  v2 encryption policies did switch to a better KDF in the kernel for deriving per-file keys, but this has nothing to do with the password-based KDF in userspace, and the fact that systemd doesn't create metadata that is compatible with the `fscrypt` tool.  Although systemd and the `fscrypt` tool use the same underlying kernel feature, they do the userspace part of key management in different ways, so there is no compatibility between the two.

Now, I'm not necessarily saying there should be compatibility -- it looks like it was necessary to do things yourself anyway?  But I'm not sure why you're claiming that systemd-homed fscrypt directories can be accessed using standard tools.
